### ISSUE TITLE
[Issue]: Flash attention APIs doesn't support customized attention masks

### ISSUE THREAD

**Issue by jiagaoxiang** [ROCm/aiter#92]
**Note: this feature request is critical to vision or multimodal models like Llama3.2-11b-vision model where non-causal masks are common.**

The PyTorch SDPA kernel (memory-efficient attention backend) on ROCm is known to be slower than CUDA, as observed across multiple projects. To benchmark against the H100, these projects require replacing the SDPA API with the Flash Attention (FA) API. However, the FA API does not support custom attention masks or biases, allowing only a boolean flag to indicate whether a causal mask should be applied. This limitation prevents its use in scenarios requiring non-causal masks, such as padding masks (illustrated in the images below, where white squares indicate padding positions), which are commonly used in vision and multimodal models.

To address this challenge, we can pursue two possible solutions:

Further optimize the SDPA memory-efficient attention backend.
Extend the FA API to support custom attention masks or biases, enabling users to modify attention weights before applying softmax.
One important consideration is that while SDPA supports the FA backend, when a custom mask is provided, it defaults to the memory-efficient attention backend.

Ideally, both approaches should be pursued to provide flexibility. However, if prioritization is required, optimizing SDPA (Option 1) should take precedence, as many PyTorch-based repositories directly use SDPA for attention mechanisms. Improving SDPA would eliminate the need for additional work to replace it with the FA API.

![Image](https://github.com/user-attachments/assets/14c908ff-b99c-466a-98cd-c3390489d177)

### GPU

MI300, MI325

**Comment by carlushuang** [comment]
@jiagaoxiang the FA/SDPA using ck-fmha underneath, please see https://github.com/ROCm/composable_kernel/tree/develop/example/ck_tile/01_fmha. If you look at this example, we have attention bias which is quite flexible, but the limitation of FA API is they do not expose the customized attention bias.

I think your question not related to the aiter team, it's more like a question regarding SDPA/CK-FA. We could have expose simliar API inside aiter (though is our original plan, expose as much feature as we can and not limited by  API), but if you still prefer the SDPA API, maybe you need talk with SDPA team about this

will close this issue

**Comment by jiagaoxiang** [comment]
Hi @carlushuang , if both SDPA and FA uses ck-fmha kernel, I found if I supply an attn_mask to SDPA, it will defaults to memory efficient attention and cannot use FA backend. But as you said, ck-fmha supports customized attention bias, is this restriction added by the SDPA API (which they can remove)?

**Comment by carlushuang** [comment]
I think maybe the SDPA API does not properly integrate it with ck-fmha attention bias mechanism. though I'm not that familiar with SDPA and how that should designed
